# EDA and Feature Generation  



In [1]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
import numpy as np
from google.colab import drive

## Importing Data

We have ad data provided by Avazu. The data was taken from a Kaggle Competition of [Click-Through Rate Prediction](https://www.kaggle.com/c/avazu-ctr-prediction/data). We use pandas to import a subset of a very large data. The original data is ~7GB in size, so we have taken the first 200000 rows for our analysis. 

In [2]:
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
cd 'drive/My Drive'

/content/drive/My Drive


In [4]:
df = pd.read_csv("data/avazu_ctr/train_subset.csv")

In [5]:
df.head()

id  click                 hour  ...  C19     C20 C21
0  10015140740686523448      0  2014-10-21 00:00:00  ...  297  100111  61
1  10070328440095985756      1  2014-10-21 00:00:00  ...   35  100084  79
2  10093977800236804132      1  2014-10-21 00:00:00  ...   35      -1  79
3  10104245282042838695      0  2014-10-21 00:00:00  ...   35  100084  79
4  10105971003478261107      0  2014-10-21 00:00:00  ...   35      -1  79

[5 rows x 24 columns]

Given above data, it is necessary we understand what each column means.

1. id: ad identifier more specifically the add ID
2. click: 0/1 for non-click/click (this is the target variable. indicates if this ad was clicked or not)
3. hour: format is YYMMDDHH, so 14091123 means 23:00 on Sept. 11, 2014 UTC. 
4. C1: anonymized categorical variable
5. banner_pos: position when the ad might be displayed
6. site_id: ID of the website
7. site_domain: domain where the site is hosted
8. site_category: category to which the site belongs
9. app_id: application ID
10. app_domain: application domain
11. app_category: application category
12. device_id: ID of the device from which the ad is clicked
13. device_ip: Network IP to which the device was connected to while clicking on the ad (eg: 192.145.86.35)
14. device_model: model of the device used for clicking the ad.
15. device_type: type of device used for clicking the ad (eg: laptop, desktop, mobile)
16. device_conn_type: connection type of the device (LAN, wifi, etc)
17. C14-C21: anonymized categorical variables (although being anonymous, variables C15 and C16 seems to give the dimensions of the ad on the page in terms of pixels)

In [6]:
df.shape

(200000, 24)

Our data has 200000 rows / records and 24 features / columns. 

In [7]:
df.dtypes

id                  uint64
click                int64
hour                object
C1                   int64
banner_pos           int64
site_id             object
site_domain         object
site_category       object
app_id              object
app_domain          object
app_category        object
device_id           object
device_ip           object
device_model        object
device_type          int64
device_conn_type     int64
C14                  int64
C15                  int64
C16                  int64
C17                  int64
C18                  int64
C19                  int64
C20                  int64
C21                  int64
dtype: object

Our features can be broadly classified into following categories:
    
1. **site features:** site_id, site_category, site_domain
2. **app features:** app_id, app_domain, app_category
3. **device features:** device_id, device_ip, device_model, device_type, device_conn_type
4. **anonymized categorical features** C1 & C14-C21
5. **other features:** hour, banner_pos
6. **target variable:** click

In [8]:
df["hour"] = pd.to_datetime(df["hour"])

## Exploring and Preprocessing Data with Feature Engineering

### 1. Checking for NaNs and missing values

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Data columns (total 24 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                200000 non-null  uint64        
 1   click             200000 non-null  int64         
 2   hour              200000 non-null  datetime64[ns]
 3   C1                200000 non-null  int64         
 4   banner_pos        200000 non-null  int64         
 5   site_id           200000 non-null  object        
 6   site_domain       200000 non-null  object        
 7   site_category     200000 non-null  object        
 8   app_id            200000 non-null  object        
 9   app_domain        200000 non-null  object        
 10  app_category      200000 non-null  object        
 11  device_id         200000 non-null  object        
 12  device_ip         200000 non-null  object        
 13  device_model      200000 non-null  object        
 14  devi

The first thing which we check usually is if there is NaN values. Our data doesn't seem to have null values and all the data seems to be there. If in case we had null values, we could have replaced them by mean of values of the column one at a time, or looking at other related columns and then adapting our missing values accordingly. 

### 2. Identify unique values for columns

In [10]:
def count_unique(d, columns):
    for column in columns:
        print("Number of Unique values in column {} is {}".format(column, str(len(d[column].unique()))))

In [11]:
columns = list(df.columns)
count_unique(df, columns)

Number of Unique values in column id is 200000
Number of Unique values in column click is 2
Number of Unique values in column hour is 240
Number of Unique values in column C1 is 7
Number of Unique values in column banner_pos is 7
Number of Unique values in column site_id is 1788
Number of Unique values in column site_domain is 1707
Number of Unique values in column site_category is 21
Number of Unique values in column app_id is 1732
Number of Unique values in column app_domain is 117
Number of Unique values in column app_category is 22
Number of Unique values in column device_id is 33169
Number of Unique values in column device_ip is 143819
Number of Unique values in column device_model is 3756
Number of Unique values in column device_type is 4
Number of Unique values in column device_conn_type is 4
Number of Unique values in column C14 is 1884
Number of Unique values in column C15 is 8
Number of Unique values in column C16 is 9
Number of Unique values in column C17 is 404
Number of Un

### 3. Exploring Click-Through Rate (CTR)

Here, we would like to know the Click-Through Rate (CTR) of the given dataset which we have. CTR is defined as the number of users who click an ad on a particular page to the total number of users who happen to visit that page. A higher CTR indicates that a lot of users were interested in our ads which we hosted on a particular website / page.
Below we try to observe how many people in our data actually clicked the ad. We then calculate the CTR to get an estimate of how well our ad is performing.

In [12]:
fig = px.histogram(df, x="click")
fig.update_layout(title="Click histogram")
fig.show()
# fig.write_html("plots/click_histogram.html")
# IFrame('plots/click_histogram.html', height=600, width=1000)

In [13]:
df["click"].value_counts()

0    165748
1     34252
Name: click, dtype: int64

In [14]:
CTR = len(df[df["click"] == 1]) / len(df)
print("Click-Through Rate (CTR): {}".format(str(CTR)))

Click-Through Rate (CTR): 0.17126


From the above histogram and statistic, we can observe that about 166132 users do not click on our ad but only a small fraction of 33868 users happen to click our ad. Our CTR is **~17%**. This means that about **83%** of the people do not click on the ad at all! 

### 4. Effects of Days and Time on Clicks

Here we explore our datetime feature **hour** and observe how CTR varies based on days and various time of the day. This is particularly important since it will help us give insights as to what kind of ads can we put and at what time of the day or week is the traffic most promising for giving us profits. We know that the amount of people doing online shopping on **Black Friday** or **Cyber Monday** is high. These days usually occur every year in October-November period. Likewise, since most of the people are working during the day, we can assume the amount of people visiting websites or clicking ads is higher during night time or any time after **5-6 pm** more specifically after the official business hours. We also need to consider weekends as well as geographic locations for which our CTR scores can be significantly impacted. Other possible events include, **election periods**, **festivals**, etc where we might expect people to click on ads more. Hence, **datetime** based analysis needs to be done to understand the CTR trend

#### 4.1 Days vs Clicks per hour

In [15]:
df["hour"].describe()

count                  200000
unique                    240
top       2014-10-22 09:00:00
freq                     2238
first     2014-10-21 00:00:00
last      2014-10-30 23:00:00
Name: hour, dtype: object

In [16]:
click_day = df.groupby('hour').agg({'click':'sum'}).reset_index()
click_day.head()

hour  click
0 2014-10-21 00:00:00    101
1 2014-10-21 01:00:00    133
2 2014-10-21 02:00:00    156
3 2014-10-21 03:00:00    189
4 2014-10-21 04:00:00    182

In [17]:
fig = go.Figure(go.Scatter(name="clicks/day",
    x = click_day['hour'],
    y = click_day['click'],
    hovertemplate='Date: %{x|%d %B %Y} <br>Time: %{x|%H:%M:%S} <br>Day: %{x|%A} <br>Clicks: %{y}'
))

fig.update_layout(
    title = 'Trend of clicks grouped by day for all hours',
    xaxis_tickformat = '%d %B <br>%Y',
    xaxis_title = "Hourly clicks for 10 days",
    yaxis_title = "Number of Clicks"
)

fig.show()

Above we have a plot of the amount of clicks made every hour for the 10 days of data given between **21st October 2014** to **31st October 2014**. We see peaks in the clicks made on **22nd** and **28th** of October somewhere around mid-day. Likewise, we see a surprising dip during **25th** of October at night. Apart from these 3 outlier peaks, the hourly click rate seems pretty stationary and the trend seems to be almost the same for the rest of the days.

#### 4.2 Hours vs Clicks 

Earlier we plotted days vs clicks done by users per hour. Now, we would like to see how many clicks were made for each hour for all the days. Basically we sum all the clicks made for the first hour of all the days, the second hour for all the days, etc for all the hours in the day. Our X-axis will be hours all the 24 hours. This will give us the trend of the how the clicks vary every day for a particular hour. We perform feature engineering to achieve our plot

In [18]:
df['hour_of_day'] = df["hour"].apply(lambda x: str(x.time())[:5])
click_hr = df.groupby('hour_of_day').agg({'click':'sum'}).reset_index()
click_hr.head()

hour_of_day  click
0       00:00    738
1       01:00    948
2       02:00   1063
3       03:00   1228
4       04:00   1463

In [19]:
fig = go.Figure(go.Scatter(name="clicks/hr",
    x = click_hr['hour_of_day'],
    y = click_hr['click'],
    hovertemplate='Time: %{x} <br>Clicks: %{y}'
))

fig.update_layout(
    title = 'Trend of clicks grouped by hours for all the days',
    xaxis_title = "Hours",
    yaxis_title = "Number of Clicks"
)

fig.show()

From the above trend the highest clicks are made every day during **12:00 pm** to **2:00 pm**. The amount of clicks done is less during the initial and the later part of the day. This means that people become more active during the business hours of the day, rather than towards the end of the day or the beginning of the day! This confirms our earlier observation as described by the earlier chart.

#### 4.3 Hourly Impressions based on clicks

**Impressions** are when ads are rendered on a user screen or any other form of digital media platform. Impressions are not action-based and are merely defined by a user potentially seeing the advertisement. Hence, it doesn't really matter if someone clicked the ad or not, the **impressions** are just the fact that the ad was observed by any person and they saw it with/without any action taken on it.

In [20]:
df.head()

id  click       hour    C1  ...  C19     C20 C21 hour_of_day
0  10015140740686523448      0 2014-10-21  1005  ...  297  100111  61       00:00
1  10070328440095985756      1 2014-10-21  1005  ...   35  100084  79       00:00
2  10093977800236804132      1 2014-10-21  1005  ...   35      -1  79       00:00
3  10104245282042838695      0 2014-10-21  1005  ...   35  100084  79       00:00
4  10105971003478261107      0 2014-10-21  1005  ...   35      -1  79       00:00

[5 rows x 25 columns]

We group our data firstly based on hour and then based on click. This helps us achieve multi-level grouping. However, we would like to bring all data to one level hence we unstack() it and plot a graph comparing clicks and non-clicks for every hour for all the days

In [21]:
impressions = df.groupby(['hour_of_day', 'click']).size().unstack().reset_index()
impressions.head()

click hour_of_day     0     1
0           00:00  3538   738
1           01:00  3952   948
2           02:00  5003  1063
3           03:00  5612  1228
4           04:00  7997  1463

In [22]:
fig = go.Figure(data=[
    go.Bar(name='Clicked', x=impressions["hour_of_day"], y=impressions[1], 
           hovertemplate='Time: %{x} <br>Clicks: %{y}', marker_color='rgb(55, 83, 109)'),
    go.Bar(name='Not Clicked', x=impressions["hour_of_day"], y=impressions[0], 
           hovertemplate='Time: %{x} <br>Clicks: %{y}', marker_color='rgb(26, 118, 255)')
])
# Change the bar mode
fig.update_layout(
    title = 'Hourly Impressions based on Clicks',
    xaxis_title = "Hour of the day",
    yaxis_title = "Impressions / hr",
    barmode='group',
    )
fig.show()

Above figure shows us hourly impressions, which means that for every hour, a significantly high number of people saw the ads but only a fraction of them actually clicked it and were forwarded to a landing page.

#### 4.4 Hourly Click-Through Rate (CTR)

Earlier we saw hourly and daily clicks made on our ads. Now we would like to observe the hourly Click-Through Rate (CTR). **Click-Through Rate** is the number of times the ad was clicked by the total impressions. We calculate how many times the ad was clicked in an hour and divide it by the total impressions of that hour. This will give us hourly Click-Through Rate. 

In [23]:
just_clicks = df[df['click'] == 1]
hourly_ctr = df[["hour_of_day", "click"]].groupby(["hour_of_day"]).count().reset_index()
hourly_ctr = hourly_ctr.rename(columns={'click': 'impressions'})
hourly_ctr["clicks"] = just_clicks[["hour_of_day", "click"]].groupby(["hour_of_day"]).count().reset_index()["click"]
hourly_ctr["CTR"] = hourly_ctr["clicks"] / hourly_ctr["impressions"] * 100
hourly_ctr.head()

hour_of_day  impressions  clicks        CTR
0       00:00         4276     738  17.259121
1       01:00         4900     948  19.346939
2       02:00         6066    1063  17.523904
3       03:00         6840    1228  17.953216
4       04:00         9460    1463  15.465116

In [24]:
fig = px.bar(hourly_ctr, x='hour_of_day', y='CTR',
             labels={"hour_of_day": "Time"}, 
             color='CTR',
             height=400)

fig.update_layout(
    title = 'Hourly CTR',
    xaxis_title = "Hour of the day",
    yaxis_title = "Click-Through Rate (CTR)",
    )
fig.show()

Contrary to what we observed earlier of the clicks being higher during the afternoon time or during the mid of the day, the CTR values suggest that a higher number of users click an ad relative to the impressions during midnight at around **1:00 am** and likewise, the second highest peaks are at **3:00 pm. - 4:00 pm** in the evening. If we just consider impressions then mid-night had less impressions relative to other times of the day and the same goes to the number of clicks done on the ad; however, considering both, it's an interesting trend to see the CTR to be high during the early time of the day.

#### 4.5 Daily CTR 

Now that we know how the CTR trend is for every hour of the day, let's observe how it is for every day in the week. We will basically observe 3 things:

1. Number of clicks made for each day i.e trend of clicks for each day
2. Number of impressions we had for each day for both the click and no-click cases
3. Daily CTR projections to observe how the trend if for each day

In [25]:
df.head()

id  click       hour    C1  ...  C19     C20 C21 hour_of_day
0  10015140740686523448      0 2014-10-21  1005  ...  297  100111  61       00:00
1  10070328440095985756      1 2014-10-21  1005  ...   35  100084  79       00:00
2  10093977800236804132      1 2014-10-21  1005  ...   35      -1  79       00:00
3  10104245282042838695      0 2014-10-21  1005  ...   35  100084  79       00:00
4  10105971003478261107      0 2014-10-21  1005  ...   35      -1  79       00:00

[5 rows x 25 columns]

In [26]:
days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
df["day_of_week"] = df["hour"].apply(lambda x: days[x.weekday()])
click_days = df.groupby("day_of_week").agg({"click": "sum"}).reset_index()
click_days['day_of_week'] = pd.Categorical(click_days['day_of_week'], categories=days, ordered=True)
click_days = click_days.sort_values('day_of_week')
click_days.head(7)

day_of_week  click
1      Monday   2922
5     Tuesday   7693
6   Wednesday   7249
4    Thursday   6967
0      Friday   2873
2    Saturday   3079
3      Sunday   3469

In [27]:
fig = go.Figure(go.Scatter(name="clicks/day",
    x = click_days['day_of_week'],
    y = click_days['click'],
    hovertemplate='Day: %{x} <br>Clicks: %{y}',
    marker_color="darkolivegreen"
))

fig.update_layout(
    title = 'Trend of clicks grouped by days',
    xaxis_title = "Day of Week",
    yaxis_title = "Number of Clicks"
)

fig.show()

In [28]:
impressions = df.groupby(['day_of_week', 'click']).size().unstack().reset_index()
impressions['day_of_week'] = pd.Categorical(impressions['day_of_week'], categories=days, ordered=True)
impressions = impressions.sort_values('day_of_week')
impressions.head(7)

click day_of_week      0     1
1          Monday  13083  2922
5         Tuesday  38951  7693
6       Wednesday  37999  7249
4        Thursday  32918  6967
0          Friday  13579  2873
2        Saturday  13605  3079
3          Sunday  15613  3469

In [29]:
fig = go.Figure(data=[
    go.Bar(name='Clicked', x=impressions["day_of_week"], y=impressions[1], 
           hovertemplate='Day: %{x} <br>Clicks: %{y}', marker_color='indianred'),
    go.Bar(name='Not Clicked', x=impressions["day_of_week"], y=impressions[0], 
           hovertemplate='Day: %{x} <br>Clicks: %{y}', marker_color='lightsalmon')
])
# Change the bar mode
fig.update_layout(
    title = 'Daily Impressions based on Clicks',
    xaxis_title = "Day of week",
    yaxis_title = "Impressions / day",
    barmode='group',
    )
fig.show()

In [30]:
just_clicks = df[df['click'] == 1]
daily_ctr = df[["day_of_week", "click"]].groupby(["day_of_week"]).count().reset_index()
daily_ctr = daily_ctr.rename(columns={'click': 'impressions'})
daily_ctr["clicks"] = just_clicks[["day_of_week", "click"]].groupby(["day_of_week"]).count().reset_index()["click"]
daily_ctr["CTR"] = daily_ctr["clicks"] / daily_ctr["impressions"] * 100
daily_ctr['day_of_week'] = pd.Categorical(daily_ctr['day_of_week'], categories=days, ordered=True)
daily_ctr = daily_ctr.sort_values('day_of_week')
daily_ctr.head(7)

day_of_week  impressions  clicks        CTR
1      Monday        16005    2922  18.256795
5     Tuesday        46644    7693  16.493011
6   Wednesday        45248    7249  16.020598
4    Thursday        39885    6967  17.467720
0      Friday        16452    2873  17.462922
2    Saturday        16684    3079  18.454807
3      Sunday        19082    3469  18.179436

In [31]:
fig = px.bar(daily_ctr, x='day_of_week', y='CTR',
             labels={"day_of_week": "Day"}, 
             color='CTR',
             height=400)

fig.update_layout(
    title = 'Daily CTR',
    xaxis_title = "Day of week",
    yaxis_title = "Click-Through Rate (CTR)",
    )
fig.show()

Our daily CTR graph shows that on Saturday and on Sunday the chances of the ad being clicked is higher. This is reasonable since on weekends people will have more time to spend online and come across ads and click them. 

Now that we have understood the effect of clicks based on hours and days in the week, with different combinations, let us understand effect of other variables on the target click.

### 5. Effect of Site features of Clicks

The kind of website hosting our ads as a huge impact on our clicks. Is the website a famous one, is it a commercial e-commerce one, or is it just a blogging site, etc; a lot of factors related to site plays an important role into whether a person will click an ad rendered on it. Firstly as calculated earlier, we have **1788** unique websites. 

#### 5.1 Effects of Site id on clicks and CTR

In [32]:
print("Number of unique websites: {}".format(str(len(df["site_id"].unique()))))

Number of unique websites: 1788


In [33]:
# top5 websites based on number of ads displayed in them
siteids = df["site_id"].value_counts()[:5].index
site_impressions = df["site_id"].value_counts()[:5].values
print("Top5 websites based on impressions: \n{}".format(siteids))

Top5 websites based on impressions: 
Index(['85f751fd', '1fbe01fe', 'e151e245', 'd9750ee7', '5b08c53b'], dtype='object')


In [34]:
top5_sites = df[(df["site_id"].isin(siteids))]
top5_sites_click = top5_sites.groupby(['site_id', 'click']).size().unstack().reset_index()
top5_sites_click = top5_sites_click.sort_values(by=1, ascending=False).reset_index()
top5_sites_click["site_impressions"] = site_impressions
top5_sites_click = top5_sites_click.rename(columns={0: 'Not Clicked', 1: "Clicked"})
top5_sites_click.columns.name = None
top5_sites_click = top5_sites_click.drop(["index"], axis=1)
top5_sites_click.head()

site_id  Not Clicked  Clicked  site_impressions
0  85f751fd        63793     8706             72499
1  1fbe01fe        25416     6749             32165
2  e151e245         9218     3901             13119
3  5b08c53b         2330     2123              4741
4  d9750ee7         3361     1380              4453

In [35]:
fig = go.Figure(data=[
    go.Bar(name='Clicked', x=top5_sites_click["site_id"], y=top5_sites_click["Clicked"],
           hovertemplate='Site ID: %{x} <br>Clicks: %{y}', marker_color='seagreen'),
    go.Bar(name='Not Clicked', x=top5_sites_click["site_id"], y=top5_sites_click["Not Clicked"], 
           hovertemplate='Site ID: %{x} <br>Clicks: %{y}', marker_color='firebrick')
])
# Change the bar mode
fig.update_layout(
    title = 'Top5 Sites based on Clicks',
    xaxis_title = "Top5 Site IDs",
    yaxis_title = "Impressions / site",
    barmode='group',
    )
fig.show()

Of the 1788 sites on which our ads are placed, we have the top 5 sites in terms of amount of impressions they had. As before, a lot of people happen to see the ads but only few of them end up clicking on them. This is evident by the green bars shown above.

In [36]:
top5_sites_click['CTR'] = top5_sites_click['Clicked'] / top5_sites_click['site_impressions'] * 100
top5_sites_click.head()

site_id  Not Clicked  Clicked  site_impressions        CTR
0  85f751fd        63793     8706             72499  12.008441
1  1fbe01fe        25416     6749             32165  20.982434
2  e151e245         9218     3901             13119  29.735498
3  5b08c53b         2330     2123              4741  44.779582
4  d9750ee7         3361     1380              4453  30.990344

In [37]:
fig = px.bar(top5_sites_click, x='site_id', y='CTR',
             labels={"site_id": "Site Id"}, 
             color='CTR',
             height=400)

fig.update_layout(
    title = 'CTR values of Top5 Sites',
    xaxis_title = "Top5 Site Ids",
    yaxis_title = "Click-Through Rate (CTR)",
    )
fig.show()

We see that although site id **85f751fd** had more impressions, site id **5b08c53b** had high CTR value. So it might be the case that this sight must be having keywords which really describe the ads and that on having clicked on it the user is directed to an appropriate landing page. 

#### 5.2 Effects of Site Domain on Clicks and CTR

In [38]:
print("Number of unique domains: {}".format(str(len(df["site_domain"].unique()))))

Number of unique domains: 1707


In [39]:
# top5 domains based on number of ads displayed in them
sitedomains = df["site_domain"].value_counts()[:5].index
domain_impressions = df["site_domain"].value_counts()[:5].values
print("Top5 site domains based on impressions: \n{}".format(sitedomains))

Top5 site domains based on impressions: 
Index(['c4e18dd6', 'f3845767', '7e091613', '7687a86e', '98572c79'], dtype='object')


In [40]:
top5_domains = df[(df["site_domain"].isin(sitedomains))]
top5_domains_click = top5_domains.groupby(['site_domain', 'click']).size().unstack().reset_index()
top5_domains_click = top5_domains_click.sort_values(by=1, ascending=False).reset_index()
top5_domains_click["domain_impressions"] = domain_impressions
top5_domains_click = top5_domains_click.rename(columns={0: 'Not Clicked', 1: "Clicked"})
top5_domains_click.columns.name = None
top5_domains_click = top5_domains_click.drop(["index"], axis=1)
top5_domains_click.head()

site_domain  Not Clicked  Clicked  domain_impressions
0    c4e18dd6        65812     9307               75119
1    f3845767        25416     6749               32165
2    7e091613        12265     4288               16553
3    7687a86e         3381     2949                6330
4    98572c79         3500     1395                4895

In [41]:
fig = go.Figure(data=[
    go.Bar(name='Clicked', x=top5_domains_click["site_domain"], y=top5_domains_click["Clicked"],
           hovertemplate='Domain ID: %{x} <br>Clicks: %{y}', marker_color='seagreen'),
    go.Bar(name='Not Clicked', x=top5_domains_click["site_domain"], y=top5_domains_click["Not Clicked"], 
           hovertemplate='Domain ID: %{x} <br>Clicks: %{y}', marker_color='firebrick')
])
# Change the bar mode
fig.update_layout(
    title = 'Top5 Domains based on Clicks',
    xaxis_title = "Top5 Site Domains",
    yaxis_title = "Impressions / domain",
    barmode='group',
    )
fig.show()

Our websites are described by a domain. If a domain is descriptive and apt then chances of people visiting it is higher, although it does not necessarily guarantee they will click the ad. If the ad is not relevant to your content or related to your core idea, it will have less CTR.

In [42]:
top5_domains_click['CTR'] = top5_domains_click['Clicked'] / top5_domains_click['domain_impressions'] * 100
top5_domains_click.head()

site_domain  Not Clicked  Clicked  domain_impressions        CTR
0    c4e18dd6        65812     9307               75119  12.389675
1    f3845767        25416     6749               32165  20.982434
2    7e091613        12265     4288               16553  25.904670
3    7687a86e         3381     2949                6330  46.587678
4    98572c79         3500     1395                4895  28.498468

In [43]:
fig = px.bar(top5_domains_click, x='site_domain', y='CTR',
             labels={"site_domain": "Domain ID"}, 
             color='CTR',
             height=400)

fig.update_layout(
    title = 'CTR values of Top5 Domains',
    xaxis_title = "Top5 Domains",
    yaxis_title = "Click-Through Rate (CTR)"
    )
fig.show()

Again, the 4th site has higher CTR although it had less impressions overall as compared to site 1

#### 5.3 Effects of Site Category on Clicks and CTR

In [44]:
print("Number of website categories: {}".format(str(len(df["site_category"].unique()))))

Number of website categories: 21


In [45]:
# top5 site categories based on number of ads displayed in them
sitecategories = df["site_category"].value_counts()[:5].index
category_impressions = df["site_category"].value_counts()[:5].values
print("Top5 site categories based on impressions: \n{}".format(sitecategories))

Top5 site categories based on impressions: 
Index(['50e219e0', 'f028772b', '28905ebd', '3e814130', 'f66779e6'], dtype='object')


In [46]:
top5_categories = df[(df["site_category"].isin(sitecategories))]
top5_categories_click = top5_categories.groupby(['site_category', 'click']).size().unstack().reset_index()
top5_categories_click = top5_categories_click.sort_values(by=1, ascending=False).reset_index()
top5_categories_click["category_impressions"] = category_impressions
top5_categories_click = top5_categories_click.rename(columns={0: 'Not Clicked', 1: "Clicked"})
top5_categories_click.columns.name = None
top5_categories_click = top5_categories_click.drop(["index"], axis=1)
top5_categories_click.head()

site_category  Not Clicked  Clicked  category_impressions
0      f028772b        51392    11330                 82064
1      50e219e0        71442    10622                 62722
2      28905ebd        28778     7714                 36492
3      3e814130        10608     4264                 14872
4      f66779e6         1111       50                  1161

In [47]:
fig = go.Figure(data=[
    go.Bar(name='Clicked', x=top5_categories_click["site_category"], y=top5_categories_click["Clicked"],
           hovertemplate='Category ID: %{x} <br>Clicks: %{y}', marker_color='seagreen'),
    go.Bar(name='Not Clicked', x=top5_categories_click["site_category"], y=top5_categories_click["Not Clicked"], 
           hovertemplate='Category ID: %{x} <br>Clicks: %{y}', marker_color='firebrick')
])
# Change the bar mode
fig.update_layout(
    title = 'Top5 Categories based on Clicks',
    xaxis_title = "Top5 Site Categories",
    yaxis_title = "Impressions / site category",
    barmode='group',
    )
fig.show()

Sites can belong to various categories - ecommerce websites, healthcare websites, education websites, etc. Each category has various websites of different domains. Above graph shows how impressions vary based on site category. For instance the 2nd site category has highest impressions. Maybe it might depict ecommerce site like Amazon or Ebay which has higher footprint then a relatively less visited website like a hospital website or maybe an educational blog site catered to a specific audience.

In [48]:
top5_categories_click['CTR'] = top5_categories_click['Clicked'] / top5_categories_click['category_impressions'] * 100
top5_categories_click.head()

site_category  Not Clicked  Clicked  category_impressions        CTR
0      f028772b        51392    11330                 82064  13.806298
1      50e219e0        71442    10622                 62722  16.935047
2      28905ebd        28778     7714                 36492  21.138880
3      3e814130        10608     4264                 14872  28.671329
4      f66779e6         1111       50                  1161   4.306632

In [49]:
fig = px.bar(top5_categories_click, x='site_category', y='CTR',
             labels={"site_category": "Site Category ID"}, 
             color='CTR',
             height=400)

fig.update_layout(
    title = 'CTR values of Top5 Site Categories',
    xaxis_title = "Top5 Site Categories",
    yaxis_title = "Click-Through Rate (CTR)"
    )
fig.show()

As before CTR values is higher for 4th site category although its impressions are lower. 

### 6. Effect of Device Features on Clicks

What type of device a user uses or the model of the device matters a lot too when it comes to predicting ads. Smartphones, tablets, smart watches are examples of various mobile device types and understanding which device is most prone to ad clicks is important too

#### 6.1 Effect of Device id and IP on Clicks

In [50]:
print("There are {} devices in the data set".format(df["device_id"].nunique()))
print("There are {} device ip in the data set".format(df["device_ip"].nunique()))

There are 33169 devices in the data set
There are 143819 device ip in the data set


In [51]:
print('The top 5 devices that have the most impressions')
print((df["device_id"].value_counts()/len(df))[0:5])

The top 5 devices that have the most impressions
a99f214a    0.824405
0f7c61dc    0.000600
c357dbff    0.000425
936e92fb    0.000335
afeffc18    0.000230
Name: device_id, dtype: float64


In [52]:
print('The top 10 devices that have the most impressions')
print((df["device_ip"].value_counts()/len(df))[0:10])

The top 10 devices that have the most impressions
6b9769f2    0.005205
431b3174    0.003385
009a7861    0.002175
d90a7774    0.002140
285aa37d    0.002130
2f323f36    0.002110
af62faf4    0.002095
930ec31d    0.002075
af9205f9    0.002045
6394f6f6    0.001985
Name: device_ip, dtype: float64


Of the 33169 unique devices about 82% impressions were there on one particular device id. That's a significant bias. The device IP seems to have a uniform distribution since the top 10 IP's show a relatively similar number of impressions.

#### 6.2 Effect of Device Type on Clicks

In [53]:
print('The impressions by device types')
print((df["device_type"].value_counts()/len(df)))

The impressions by device types
1    0.92268
0    0.05497
4    0.01907
5    0.00328
Name: device_type, dtype: float64


In [54]:
device_clicks = df[["device_type", "click"]].groupby(["device_type", "click"]).size().unstack().reset_index()
device_clicks.columns.name = None
device_clicks = device_clicks.rename(columns={0: 'Not Clicked', 1: "Clicked"})
device_clicks.head()

device_type  Not Clicked  Clicked
0            0         8703     2291
1            1       153013    31523
2            4         3439      375
3            5          593       63

In [55]:
fig = go.Figure(data=[
    go.Bar(name='Clicked', x=device_clicks["device_type"], y=device_clicks["Clicked"],
           hovertemplate='Device Type: %{x} <br>Clicks: %{y}', marker_color='peachpuff'),
    go.Bar(name='Not Clicked', x=device_clicks["device_type"], y=device_clicks["Not Clicked"], 
           hovertemplate='Device Type: %{x} <br>Clicks: %{y}', marker_color='peru')
])

# we mention xaxis category type since it will consider it as a serial numeric data and plot missing values also like 2 and 3 which are not there in the device_type column
fig.update_layout(
    title = 'Device Types based on Clicks',
    xaxis_title = "Type of Device",
    yaxis_title = "Impressions / device type",
    xaxis_type = "category",
    barmode='group',
    )
fig.show()

In [56]:
device1 = df[df["device_type"] == 1].groupby(["hour_of_day", "click"]).size().unstack().reset_index()
device1.columns.name = None
device1 = device1.rename(columns={0: 'Not Clicked', 1: "Clicked"})
device1.head()

hour_of_day  Not Clicked  Clicked
0       00:00         3283      672
1       01:00         3624      866
2       02:00         4293      942
3       03:00         5203     1130
4       04:00         7465     1332

In [57]:
fig = px.bar(device1, x='hour_of_day', y='Clicked',
             labels={"hour_of_day": "Time"}, 
             color='Clicked',
             height=400)

fig.update_layout(
    title = 'Device type 1 hourly click trend',
    xaxis_title = "Hour of the day",
    yaxis_title = "Clicks",
    )
fig.show()

In [58]:
device_clicks["impressions"] = device_clicks["Clicked"] + device_clicks["Not Clicked"]
device_clicks["CTR"] = device_clicks["Clicked"] / device_clicks["impressions"] * 100
device_clicks.head()

device_type  Not Clicked  Clicked  impressions        CTR
0            0         8703     2291        10994  20.838639
1            1       153013    31523       184536  17.082304
2            4         3439      375         3814   9.832197
3            5          593       63          656   9.603659

In [59]:
fig = px.bar(device_clicks, x='device_type', y='CTR',
             labels={"device_type": "Device Type"}, 
             color='CTR',
             height=400)

fig.update_layout(
    title = 'CTR values for devices of different types',
    xaxis_title = "Device Type",
    yaxis_title = "Click-Through Rate (CTR)",
    xaxis_type = "category"
    )
fig.show()

We observed earlier that Device Type 1 had the maximum impressions but if we considered the CTR score then device 0 had the highest of them. 

### 7. Effect of App Features on Clicks

Which app category generated higher impression and clicks is more important. Say a user uses a gaming app, or an ecommerce app like Amazon or say other streaming service app like Netflix, then the we need to observe and understand which category - gaming, ecommerce, entertainment, health, finance, etc apps renders ads and how many of them are clicked. We have **app_id** as well as **app_domain** as other features and they won't provide much intuition about clicks as much as a category can provide.

#### 7.1 Effect of App Category on Clicks

In [60]:
print("There are {} app categories in the data set".format(df["app_category"].nunique()))

There are 22 app categories in the data set


In [61]:
print('The percentage of impressions by app categories')
app_categories = df["app_category"].value_counts().rename_axis('app_categories').reset_index(name='counts')
app_categories["counts"] = app_categories["counts"] / len(df)
app_categories.head()

The percentage of impressions by app categories


app_categories   counts
0       07d7df22  0.64590
1       0f2161f8  0.23720
2       cef3e649  0.04257
3       8ded1f7a  0.03651
4       f95efa07  0.02889

In [62]:
fig = px.bar(app_categories, x='app_categories', y='counts',
             labels={"app_categories": "App Categories"}, 
             color='counts',
             height=400)

fig.update_layout(
    title = 'Percentage Impressions for all App Categories',
    xaxis_title = "App Categories",
    yaxis_title = "Counts",
    )
fig.show()

Only 5 app categories seem to have a significant impression score. Other app categories hardly contributed to overall ad impressions. Let observe the click trend for these 5 app categories

In [63]:
top5_app = (df["app_category"].value_counts())[:5].index
top5_impressions = (df["app_category"].value_counts())[:5].values
print(top5_app, top5_impressions)

Index(['07d7df22', '0f2161f8', 'cef3e649', '8ded1f7a', 'f95efa07'], dtype='object') [129180  47440   8514   7302   5778]


In [64]:
top5_apps = df[(df["app_category"].isin(top5_app))]
top5_app_click = top5_apps.groupby(['app_category', 'click']).size().unstack().reset_index()
top5_app_click = top5_app_click.sort_values(by=0, ascending=False).reset_index()
top5_app_click["app_impressions"] = top5_impressions
top5_app_click = top5_app_click.rename(columns={0: 'Not Clicked', 1: "Clicked"})
top5_app_click.columns.name = None
top5_app_click = top5_app_click.drop(["index"], axis=1)
top5_app_click.head()

app_category  Not Clicked  Clicked  app_impressions
0     07d7df22       103236    25944           129180
1     0f2161f8        42259     5181            47440
2     cef3e649         7723      791             8514
3     8ded1f7a         6647      655             7302
4     f95efa07         4302     1476             5778

In [65]:
fig = go.Figure(data=[
    go.Bar(name='Clicked', x=top5_app_click["app_category"], y=top5_app_click["Clicked"],
           hovertemplate='App Category: %{x} <br>Clicks: %{y}', marker_color='crimson'),
    go.Bar(name='Not Clicked', x=top5_app_click["app_category"], y=top5_app_click["Not Clicked"], 
           hovertemplate='App Category: %{x} <br>Clicks: %{y}', marker_color='silver')
])

# we mention xaxis category type since it will consider it as a serial numeric data and plot missing values also like 2 and 3 which are not there in the device_type column
fig.update_layout(
    title = 'App Category based on Clicks',
    xaxis_title = "App Category",
    yaxis_title = "Impressions / app category",
    barmode='group',
    )
fig.show()

In [66]:
top5_app_click["CTR"] = top5_app_click["Clicked"] / top5_app_click["app_impressions"]
top5_app_click.head()

app_category  Not Clicked  Clicked  app_impressions       CTR
0     07d7df22       103236    25944           129180  0.200836
1     0f2161f8        42259     5181            47440  0.109212
2     cef3e649         7723      791             8514  0.092906
3     8ded1f7a         6647      655             7302  0.089701
4     f95efa07         4302     1476             5778  0.255452

In [67]:
fig = px.bar(top5_app_click, x='app_category', y='CTR',
             labels={"app_category": "App Category"}, 
             color='CTR',
             height=400)

fig.update_layout(
    title = 'CTR values for apps of different categories',
    xaxis_title = "App Category",
    yaxis_title = "Click-Through Rate (CTR)"
    )
fig.show()

Although the impressions were higher for the first app cateogory, the 5th category has a higher CTR score

### 8. Effects of Banner Position on Clicks

As we have calculated earlier, there are 7 unique banner positions where our ads can be displayed. These are positions where our ads are displayed on a webpage. Ads can be on the sides or at the very bottom, or as a pop-up, etc. 

In [68]:
print("Number of unique banner positions: {}".format(df["banner_pos"].nunique()))

Number of unique banner positions: 7


In [69]:
banner_pos = df["banner_pos"].value_counts().rename_axis('banner_pos').reset_index(name='counts')
banner_pos = banner_pos.sort_values(by="banner_pos", ascending=True).reset_index()
banner_pos = banner_pos.drop(["index"], axis=1)
banner_pos

banner_pos  counts
0           0  143899
1           1   55762
2           2      65
3           3      14
4           4      38
5           5      22
6           7     200

In [70]:
fig = px.bar(banner_pos, x='banner_pos', y='counts',
             labels={"banner_pos": "Banner Position"}, 
             color='counts',
             height=400)

fig.update_layout(
    title = 'Banner Position Histogram',
    xaxis_title = "Banner Position",
    yaxis_title = "Counts"
    )
fig.show()

The first 2 banner locations / positions were mostly preferred for displaying ads on a website. The remaining 5 had less relative impressions. However, as we have observed earlier, we would still like to see how many clicks were achieved on all banner positions since higher impressions does not guarantee higher CTR.

In [71]:
banner_pos_click = df[["banner_pos", "click"]].groupby(["banner_pos","click"]).size().unstack().reset_index()
banner_pos_click["impressions"] = banner_pos["counts"]
banner_pos_click = banner_pos_click.rename(columns={0: 'Not Clicked', 1: "Clicked"})
banner_pos_click.columns.name = None
banner_pos_click

banner_pos  Not Clicked  Clicked  impressions
0           0       120052    23847       143899
1           1        45441    10321        55762
2           2           58        7           65
3           3           10        4           14
4           4           36        2           38
5           5           21        1           22
6           7          130       70          200

In [72]:
fig = go.Figure(data=[
    go.Bar(name='Clicked', x=banner_pos_click["banner_pos"], y=banner_pos_click["Clicked"],
           hovertemplate='Banner Position: %{x} <br>Clicks: %{y}', marker_color='darkviolet'),
    go.Bar(name='Not Clicked', x=banner_pos_click["banner_pos"], y=banner_pos_click["Not Clicked"], 
           hovertemplate='Banner Position: %{x} <br>Clicks: %{y}', marker_color='deeppink')
])

# we mention xaxis category type since it will consider it as a serial numeric data and plot missing values also like 2 and 3 which are not there in the device_type column
fig.update_layout(
    title = 'Banner positions based on Clicks',
    xaxis_title = "Banner Positions",
    yaxis_title = "Impressions / banner position",
    barmode='group',
    )
fig.show()

In [73]:
banner_pos_click["CTR"] = banner_pos_click["Clicked"] / banner_pos_click["impressions"]
banner_pos_click

banner_pos  Not Clicked  Clicked  impressions       CTR
0           0       120052    23847       143899  0.165720
1           1        45441    10321        55762  0.185090
2           2           58        7           65  0.107692
3           3           10        4           14  0.285714
4           4           36        2           38  0.052632
5           5           21        1           22  0.045455
6           7          130       70          200  0.350000

In [74]:
fig = px.bar(banner_pos_click, x='banner_pos', y='CTR',
             labels={"banner_pos": "Banner Position"}, 
             color='CTR',
             height=400)

fig.update_layout(
    title = 'CTR based on Banner Positions',
    xaxis_title = "Banner Position",
    yaxis_title = "Click-Through Rate (CTR)"
    )
fig.show()

As expected, banner position 7 has the highest CTR although it had next to None the number of impressions compared to first 2 banner positions. Banner positions is therefore an important variable and it can help us understand which ad position on the website receives the highest ad impressions and CTR. 

### 9. Effect of Anonymous Features C1 and C14-C21 on Clicks

We have someone anonymous features which we don't know what exactly do they represent in the ad, however we can use them and study their effects on the click behaviour

In [75]:
def find_number_of_uniques(feature):
  print("Number of unique {} values: {}".format(feature, df[feature].nunique()))

In [76]:
def feature_impressions(feature):
  c = df[feature].value_counts().rename_axis(feature).reset_index(name='counts')
  c = c.sort_values(by=feature, ascending=True).reset_index()
  c = c.drop(["index"], axis=1)
  c[feature + "_percent"] = c["counts"] / sum(c["counts"])
  return c

In [77]:
def plot_singeBar(dframe, xax, yax, title_name, x_name, y_name):

  fig = px.bar(dframe, x=xax, y=yax,
              color=yax,
              height=400)

  fig.update_layout(
      title = title_name,
      xaxis_title = x_name,
      yaxis_title = y_name,
      xaxis_type = "category"
      )
  fig.show()

In [78]:
def calculate_clicks(feature, total_counts):
  c_click = df[[feature, "click"]].groupby([feature,"click"]).size().unstack().reset_index()
  c_click["impressions"] = total_counts  #c1["counts"]
  c_click = c_click.rename(columns={0: 'Not Clicked', 1: "Clicked"})
  c_click.columns.name = None
  c_click["Clicked"] = c_click["Clicked"].fillna(0)
  c_click["CTR"] = c_click["Clicked"] / c_click["impressions"]
  return c_click

In [79]:
def plot_groupBars_for_clicks(dframe,feature):

  title_name = feature + " values based on clicks"
  xaxis_name = feature + " values"
  yaxis_name = "Impressions / " + feature + " values"
  fig = go.Figure(data=[
      go.Bar(name='Clicked', x=dframe[feature], y=dframe["Clicked"],
            hovertemplate=feature + ': %{x} <br>Clicks: %{y}', marker_color='darkviolet'),
      go.Bar(name='Not Clicked', x=dframe[feature], y=dframe["Not Clicked"], 
            hovertemplate=feature + ': %{x} <br>Clicks: %{y}', marker_color='deeppink')
  ])

  # we mention xaxis category type since it will consider it as a serial numeric data and plot missing values also like 2 and 3 which are not there in the device_type column
  fig.update_layout(
      title = title_name,
      xaxis_title = xaxis_name,
      yaxis_title = yaxis_name,
      barmode='group',
      xaxis_type = "category"
      )
  fig.show()

#### 9.1 Effect of C1 on Clicks

In [80]:
find_number_of_uniques("C1")

Number of unique C1 values: 7


In [81]:
c1 = feature_impressions("C1")
c1

C1  counts  C1_percent
0  1001      29    0.000145
1  1002   10994    0.054970
2  1005  183810    0.919050
3  1007     147    0.000735
4  1008      22    0.000110
5  1010    4470    0.022350
6  1012     528    0.002640

In [82]:
title_name = "C1 Histogram"
x_name = "C1 values"
y_name = "Percentage of C1 values"
plot_singeBar(c1, "C1", "C1_percent", title_name, x_name, y_name)

C1 value of 1005 has 91% contribution of all the C1 values. Naturally we can assume that it will have higher clicks and non-clicks values associated with it.

In [83]:
c1_click = calculate_clicks("C1", c1["counts"])
c1_click

C1  Not Clicked  Clicked  impressions       CTR
0  1001         29.0      0.0           29  0.000000
1  1002       8703.0   2291.0        10994  0.208386
2  1005     152384.0  31426.0       183810  0.170970
3  1007        143.0      4.0          147  0.027211
4  1008         21.0      1.0           22  0.045455
5  1010       4032.0    438.0         4470  0.097987
6  1012        436.0     92.0          528  0.174242

In [84]:
plot_groupBars_for_clicks(c1_click, "C1")

In [85]:
title_name = "CTR based on C1"
x_name = "C1 values"
y_name = "Click-Through Rate (CTR)"
plot_singeBar(c1_click, "C1", "CTR", title_name, x_name, y_name)

In [86]:
print("Mean CTR for the entire dataset: {}".format(df["click"].mean()))

Mean CTR for the entire dataset: 0.17126


In [87]:
c1_click["CTR"].describe()

count    7.000000
mean     0.103464
std      0.082148
min      0.000000
25%      0.036333
50%      0.097987
75%      0.172606
max      0.208386
Name: CTR, dtype: float64

If we observe the above C1 statistic and the average CTR over entire dataset, we can observe that C1 values **1002**, **1005** and **1012** have CTR values higher than the average CTR of the dataset - 0.20, 0.170 and 0.172 respectively. Hence they contribute much more than other values

In [88]:
# number of unique values for columns c14-c21
find_number_of_uniques("C14")
find_number_of_uniques("C15")
find_number_of_uniques("C16")
find_number_of_uniques("C17")
find_number_of_uniques("C18")
find_number_of_uniques("C19")
find_number_of_uniques("C20")
find_number_of_uniques("C21")

Number of unique C14 values: 1884
Number of unique C15 values: 8
Number of unique C16 values: 9
Number of unique C17 values: 404
Number of unique C18 values: 4
Number of unique C19 values: 66
Number of unique C20 values: 157
Number of unique C21 values: 60


Let's explore C15 and C16 only since we suspect that they might represent ad resolutions based on initial data exploration

#### 9.2 Effect of feature C15 on clicks

In [89]:
c15 = feature_impressions("C15")
c15

C15  counts  C15_percent
0   120      13     0.000065
1   216    1461     0.007305
2   300   11396     0.056980
3   320  186760     0.933800
4   480       7     0.000035
5   728     344     0.001720
6   768       8     0.000040
7  1024      11     0.000055

In [90]:
title_name = "C15 values Histogram"
x_name = "C15 values"
y_name = "Percentage contribution of C15 values"
plot_singeBar(c15, "C15", "C15_percent", title_name, x_name, y_name)

The C15 values seems to resemble ad resolution on the website more specifically the width of the ad creative. The values resemble standard resolution width values of 320, 480, 728, 768, etc. If C15 represents ad width resolution, some other variable might represent height of the ad creative rendered on the website

In [91]:
c15_click = calculate_clicks("C15", c15["counts"])
c15_click

C15  Not Clicked  Clicked  impressions       CTR
0   120         13.0      0.0           13  0.000000
1   216       1265.0    196.0         1461  0.134155
2   300       7201.0   4195.0        11396  0.368112
3   320     156925.0  29835.0       186760  0.159750
4   480          5.0      2.0            7  0.285714
5   728        326.0     18.0          344  0.052326
6   768          5.0      3.0            8  0.375000
7  1024          8.0      3.0           11  0.272727

In [92]:
plot_groupBars_for_clicks(c15_click, "C15")

Looks like ad creatives which are having resolution width of 320 has higher impressions

In [93]:
title_name = "CTR based on C15"
x_name = "C15 values"
y_name = "Click-Through Rate (CTR)"
plot_singeBar(c15_click, "C15", "CTR", title_name, x_name, y_name)

The CTR for ads which are 300, 480, 768 and 1024 in width is higher. This means that probably ads which span entire webpage or screen frame or relatively smaller size ads positioned strategically are most likely to be clicked

#### 9.3 Effect of C16 on Clicks

In [94]:
c16 = feature_impressions("C16")
c16

C16  counts  C16_percent
0    20      13     0.000065
1    36    1461     0.007305
2    50  188824     0.944120
3    90     344     0.001720
4   250    8824     0.044120
5   320       7     0.000035
6   480     508     0.002540
7   768      11     0.000055
8  1024       8     0.000040

In [95]:
title_name = "C16 values Histogram"
x_name = "C16 values"
y_name = "Percentage contribution of C16 values"
plot_singeBar(c16, "C16", "C16_percent", title_name, x_name, y_name)

As suspected earlier, C16 represents height of the ad creative. 

In [96]:
c16_click = calculate_clicks("C16", c16["counts"])
c16_click

C16  Not Clicked  Clicked  impressions       CTR
0    20         13.0      0.0           13  0.000000
1    36       1265.0    196.0         1461  0.134155
2    50     158715.0  30109.0       188824  0.159455
3    90        326.0     18.0          344  0.052326
4   250       5022.0   3802.0         8824  0.430870
5   320          5.0      2.0            7  0.285714
6   480        389.0    119.0          508  0.234252
7   768          8.0      3.0           11  0.272727
8  1024          5.0      3.0            8  0.375000

In [97]:
plot_groupBars_for_clicks(c16_click, "C16")

Recall that a width of 320 in C15 feature had higher impressions while in C16 a height of 50 seems to give higher impressions. With this observation, maybe ads which are 320x50 in resolution rendered on the websites have higher impressions. 

In [98]:
title_name = "CTR based on C16"
x_name = "C16 values"
y_name = "Click-Through Rate (CTR)"
plot_singeBar(c16_click, "C16", "CTR", title_name, x_name, y_name)

Ad creatives with heights 250 or greater has higher CTR score. Again, we are assuming that C15 represents ad creative width and C16 represents ad creative height. The values of these variables strongly resemble standard resolution values. 

In [99]:
# saving CSV file for feature engineering and model prediction
df.to_csv('avazu_train.csv', index=False)
!cp avazu_train.csv "data/avazu_ctr/"

In [100]:
cd "data/avazu_ctr"

/content/drive/My Drive/data/avazu_ctr


In [101]:
ls

avazu_train.csv  test.csv  train_subset.csv
